### PR TITLE
OADP-609: Add VSR default class checks w/ refactor for VSB/VSR

### PR DESCRIPTION
**Pull request by eemcmullan**

### CODE REVIEW

File: controllers/validator.go [hunk 1]
```diff
@@ -127,21 +127,47 @@ func (r *VolumeSnapshotBackupReconciler) checkForOneDefaultStorageClass(log logr
 		return false, err
```
Comment (savitharaghunathan) [inline controllers/validator.go:127]: nit: in comment volumeSnapshotClasses -> storageclasses

File: controllers/validator.go [hunk 2]
```diff
@@ -127,21 +127,47 @@ func (r *VolumeSnapshotBackupReconciler) checkForOneDefaultStorageClass(log logr
...
+	}
+
+	return true, nil
+}
+
+func (r *VolumeSnapshotRestoreReconciler) checkForOneDefaultVSRStorageClass(log logr.Logger) (bool, error) {
+	storageClassList := storagev1.StorageClassList{}
+	storageClassOptions := []client.ListOption{}
+
+	// get all volumeSnapshotClasses in cluster
```
Comment (savitharaghunathan) [inline controllers/validator.go:162]: volumeSnapshotClasses -> storageclasses

File: controllers/common.go [hunk 3]
```diff
@@ -149,3 +152,45 @@ func checkByteArrayIsEmpty(val []byte) bool {
...
+		if boolIsDefault {
+			numDefaultClasses++
+		}
+
+		if numDefaultClasses > 1 {
+			return false, errors.New("cannot have more than one default volumeSnapshotClass")
+		}
+	}
+
+	return true, nil
```
Comment (kaovilai) [inline controllers/common.go:174]:
```suggestion
	foundDefaultClass := false
	for _, vsClass := range vsClassList.Items {

		isDefaultClass, _ := vsClass.Annotations[volumeSnapshotClassDefaultKey]
		boolIsDefault, _ := strconv.ParseBool(isDefaultClass)

		// found a default volumeSnapshotClass
		if boolIsDefault {
			if foundDefaultClass { // found more than once
				return false, errors.New("cannot have more than one default volumeSnapshotClass")
			}
			foundDefaultClass = true
		}
	}

	return true, nil
```

**Review by kaovilai — CHANGES_REQUESTED** [review]

File: controllers/common.go (hunk 3 above)
Comment (kaovilai) [reply controllers/common.go:174]: We don't care about count.. just if we found more than once.

Comment (kaovilai) [reply controllers/common.go:174]: Same thing for the storageclass

**Review by kaovilai — APPROVED** [review]

File: controllers/common.go [hunk 4]
```diff
@@ -149,3 +152,47 @@ func checkByteArrayIsEmpty(val []byte) bool {
 
 	return len(val) != 0
 }
+
+func checkForOneDefaultSnapClass(vsClassList *snapv1.VolumeSnapshotClassList, log logr.Logger) (bool, error) {
```
Comment (kaovilai) [inline controllers/common.go:156]: non blocking NIT. We are not using log here.. I would rather have this removed from param. Also error is kind of redundant. You already know if you have more than one default snap class by receiving false return value.

There isn't any complex data in the error created here.

Maybe create error in the validator.go instead.

Comment (kaovilai) [reply controllers/common.go:156]:
```suggestion
func hasOneDefaultSnapClass(vsClassList *snapv1.VolumeSnapshotClassList) bool {
```

**Pull request merged**